infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java

```diff
@@ -219,29 +245,44 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
     /**
      * Resend the messages in the queue if they have timed out.
      */
-    public LogReplicationEntryMsg resend() {
+    public LogReplicationEntryMsg resend(boolean isLogEntry) {
         LogReplicationEntryMsg ack = null;
-        boolean force = false;
+        boolean error = false;
+
         try {
             ack = processAcks();
         } catch (TimeoutException te) {
             // Exceptions thrown directly from the CompletableFuture.anyOf(cfs)
             log.warn("Caught a timeout exception while processing ACKs", te);
-            force = true;
+            error = true;
         } catch (ExecutionException ee) {
             // Exceptions thrown from the send message completable future will be wrapped around ExecutionException
             log.warn("Caught an execution exception while processing ACKs", ee);
             final Throwable cause = ee.getCause();
             if (cause instanceof TimeoutException) {
-                force = true;
+                error = true;
             }
         } catch (Exception e) {
             log.warn("Caught an exception while processing ACKs.", e);
+        } finally {
+            if (isLogEntry) {
+                if (error) {
+                    activateOrIncreaseBackpressure();
+                } else {
+                    deactivateBackpressure();
+                }
+            }
         }
 
         for (int i = 0; i < pendingMessages.getSize(); i++) {
             LogReplicationPendingEntry entry = pendingMessages.getPendingEntries().get(i);
-            if (entry.timeout(msgTimer) || force) {
+            if (entry.timeout(msgTimer) || error) {
+                // Upon an ACK failing due to a timeout, we add a linearly increasing wait between retries as
+                // a backpressure mechanism.
+                if (isLogEntry && isBackpressureActive) {
```

Comment:
rethinking the placement: I think currently the number of inflight msgs is 1, so this change works
But to be future proof, should we move this logic just before the for loop? 
I am thinking of this scenario: Say we allow 5 msgs be in flight, should we wait for sometime and send all the 5 msgs without a delay, or do we want a delay before sending each msg? 
I am inclining towards the former apprach as the latter will increase the latency even when not necessary, but would like to know yours/reviewers thoughts.

Reply:
In the case that we saw there were 2 entires, asymmetric in size with one being around 20mb and the other 8mb. Presumably the 20mb one resulted in a long running apply leading to the resends. If we go with the approach to send them in bursts, a flight of messages is likely to be resent in the event that one of them does not complete within that 5sec allotted by the enclosing future, since all the entries will be sent at roughly the same time. My thinking was that by distributing the wait across entries, you now have the added benefit of the futures expiring sequentially.

Reply:
I was under the impression that we decided with the approach which adds a backoff before each message.

I am open to the other approach also as this change does have tuning which will be done incrementally.

Reply:
Yes, the current implementation has a wait before each message